pkg/controllers/federatedcluster/clusterjoin.go

```diff
@@ -80,232 +75,162 @@ const (
-			eventRecorder.Eventf(
-				cluster,
-				corev1.EventTypeWarning,
-				EventReasonJoinClusterTimeoutExceeded,
-				"get token for cluster %q timed out",
-				cluster.Name,
-			)
-
-			currentTime := metav1.Now()
-			newCondition := &fedcorev1a1.ClusterCondition{
-				Type:               fedcorev1a1.ClusterJoined,
-				Status:             corev1.ConditionFalse,
-				Reason:             pointer.String(JoinTimeoutExceededReason),
-				Message:            pointer.String(fmt.Sprintf(JoinTimeoutExceededMessageTemplate, *joinedCondition.Message)),
-				LastProbeTime:      currentTime,
-				LastTransitionTime: &currentTime,
-			}
-
-			setClusterCondition(&cluster.Status, newCondition)
-
-			var updateErr error
-			if cluster, updateErr = client.CoreV1alpha1().FederatedClusters().UpdateStatus(
-				context.TODO(),
-				cluster,
-				metav1.UpdateOptions{},
-			); updateErr != nil {
-				return cluster, fmt.Errorf("failed to update cluster status after join timeout: %w", updateErr)
-			}
-
-			return cluster, nil
-		}
-
-		logger.Info("Continue cluster join")
-	}
-
-	// 2. The remaining steps require a cluster kube client, attempt to create one
-
-	restConfig := &rest.Config{Host: cluster.Spec.APIEndpoint}
-
-	clusterSecretName := cluster.Spec.SecretRef.Name
-	if clusterSecretName == "" {
+	if joinedCondition != nil &&
+		joinedCondition.Status == corev1.ConditionFalse &&
+		time.Since(joinedCondition.LastTransitionTime.Time) > clusterJoinTimeout {
+		// join timed out
+		logger.Error(nil, "Cluster join timed out")
 		eventRecorder.Eventf(
 			cluster,
 			corev1.EventTypeWarning,
-			EventReasonJoinClusterError,
-			"cluster %q secret is not set",
-			cluster.Name,
+			EventReasonJoinClusterTimeoutExceeded,
+			"Cluster join timed out",
 		)
-		return cluster, fmt.Errorf("cluster secret is not set")
+		return cluster, &fedcorev1a1.ClusterCondition{
+			Status:  corev1.ConditionFalse,
+			Reason:  JoinTimeoutExceededReason,
+			Message: fmt.Sprintf(JoinTimeoutExceededMessageTemplate, joinedCondition.Message),
+		}, nil
 	}
-	clusterSecret, err := kubeClient.CoreV1().Secrets(fedSystemNamespace).Get(context.TODO(), clusterSecretName, metav1.GetOptions{})
-	if err != nil && apierrors.IsNotFound(err) {
+
+	// 2. The remaining steps require a cluster kube client, attempt to create one
+
+	_, clusterKubeClient, err := getClusterClient(ctx, kubeClient, fedSystemNamespace, cluster)
+	if err != nil {
+		logger.Error(err, "Failed to create cluster client")
+		msg := fmt.Sprintf("Failed to create cluster client: %v", err.Error())
 		eventRecorder.Eventf(
 			cluster,
-			corev1.EventTypeWarning,
-			EventReasonJoinClusterError,
-			"cluster %q secret %q not found",
-			cluster.Name,
-			clusterSecretName,
+			corev1.EventTypeWarning, EventReasonJoinClusterError, msg,
 		)
-		return cluster, fmt.Errorf("cluster secret not found: %w", err)
+		return cluster, &fedcorev1a1.ClusterCondition{
+			Status:  corev1.ConditionFalse,
+			Reason:  TokenNotObtainedReason,
```

Comment:
should the reason here be `TokenNotObtained`? `TokenNotObtained` refers to not being able to find the service account token in the member cluster right? would it be misleading to reuse the same reason?